File: Ex_3.1_Decay.py
import ROOT
from math import *

#def Likelyhood(x, theta):
#  likely = 1.
#  for xi in x:
#    likely = likely * xi * theta
#    print likely
#  return likely


n_bin = 100
N = 10000

r = ROOT.TRandom3(0)
c = ROOT.TCanvas("c","c",800,600)
h = ROOT.TH1F("h","h", n_bin, 0 , 10)

tau = 1.

for i in range(N):
  tmp = r.Rndm()
  x = log(tau/tmp)
  h.Fill(x)
  
x=[]

h.Draw('hist')

#for b in range(n_bin):
#  x.append(h.GetBinContent(b))
#print x
  
#print Likelyhood(x, 1.)


h2 = ROOT.TH1F("h2", "h2", n_bin, 0, 2)

for i in range(1000):
 t=1.
 for k in range(10):
  tmp = r.Rndm()
  t = t + log(tau/tmp)
 h2.Fill(t/10.)

c2 = ROOT.TCanvas("c2","c2",800,600)
h2.Draw('hist')

c3 = ROOT.TCanvas("c3","c3",800,600)
h3 = ROOT.TH1F("h3", "h3", n_bin, 0, 2)
h4 = ROOT.TH1F("h4", "h4", n_bin, 0, 2)
h5 = ROOT.TH1F("h5", "h5", n_bin, 0, 2)

lam=1.   
num=[5,10,100]
for i in range(1000):
 for n in num:
  l = 1.
  for k in range(n):
   tmp = r.Rndm()
   l = l + log(lam*tau)
  if n==num[0]: 
   h3.Fill(l/n)
  elif n==num[1]:
   h4.Fill(l/n)
  elif n==num[2]:
   h5.Fill(l/n)

h3.Draw('hist')
h4.SetLineColor(ROOT.kBlue)
h4.Draw('same hist')
h5.SetLineColor(ROOT.kGreen)
h5.Draw('same hist')

close = raw_input()

File: Ex_3.1_Decay.cxx
//Ex_3.1_Decay


double likelyhood(double* x, double theta){
  double likely = 1.;
  int i=0;
  do {
    likely = likely * x[i] * theta;
    cout << likely << endl;
    i++;
  }
  while(x[i]!=x[-1]);
  return likely;
}

double max_likelyhood(double* x, double theta_min, double theta_max, int steps){
  double besttheta = 0;
  double bestlikely = 0;
  double l = 0;
  for (double i = theta_min; i<theta_max; i+((theta_max-theta_min)/steps)){
    l = likelyhood(x,i);
    if (l>bestlikely){
      bestlikely = l;
      besttheta = i;
    }
  }
  return besttheta;
}
  

void expdecay(){
  //generate 10000 Randomnumbers and apply transformation method
  const int nbins = 100;
  TRandom3* r = new TRandom3(0);
  TCanvas* c = new TCanvas("c","c",800,600);
  TH1F* h = new TH1F("h","h", nbins, 0 , 10);
  //T1F* L = new TF1("L", "L", nbins, 0, 10);
  const int N = 10000;
  double tau = 1.;
//   double f = 0.;
//   double likelyhood(double* likely, double* x, double* par){
//     double likely = 1.;
//     //do {
//     //  likely = likely * x[i] * par[0];}
//     //while(x[i]!=x[-1]);
//     return likely * x[i] * par[0];}
//   double sum = 0;
  
  for( int i = 0; i < N; i++ ){
    double tmp = r->Rndm();
    double x = log(tau/tmp);
    h->Fill(x);
  }
  double norm = 1./(h->Integral());
  //h->Scale(norm);
  //h->Fit("likelyhood");
  //h->Fit("l");
  //h->Draw("HIST");
  
  double x[nbins];
  for (int i = 0; i<nbins; i++){
    x[i] = h->GetBinContent(i);
    cout << x[i] << endl;
  }
  
  double besttheta = 0;
  besttheta = max_likelyhood(&x, 0.9, 1.1, 20);
  
  cout << "Likelyhood Fit = " << besttheta << endl;
//   double like = 0.;
//   like = likelyhood(&x,1.);
//   cout << "Likelyhood Fit = " << like << endl;



  TH1F* h1 = new TH1F("h1","h1", 100, 0 , 2);

  
  c->Close();
  
  for (int i=0; i<1000; i++){
    for (int k=0; k<10; k++){
      double tmp = r->Rndm();
      double x = log(tau/tmp);
      h1->Fill(x); 
    }
  }
}